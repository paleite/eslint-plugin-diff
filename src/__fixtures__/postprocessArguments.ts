import type { Linter } from "eslint";

const postprocessArguments: [
  [[Linter.LintMessage, ...Linter.LintMessage[]], ...Linter.LintMessage[][]],
  string
] = [
  [
    [
      {
        ruleId: "curly",
        severity: 1,
        message: "Expected { after 'if' condition.",
        line: 1,
        column: 1,
        nodeType: "IfStatement",
        messageId: "missingCurlyAfterCondition",
      },
      {
        ruleId: "curly",
        severity: 1,
        message: "Expected { after 'if' condition.",
        line: 2,
        column: 1,
        nodeType: "IfStatement",
        messageId: "missingCurlyAfterCondition",
      },
      {
        ruleId: "curly",
        severity: 1,
        message: "Expected { after 'if' condition.",
        line: 3,
        column: 1,
        nodeType: "IfStatement",
        messageId: "missingCurlyAfterCondition",
        fix: {
          range: [0, 3],
          text: ''
        }
      },
    ],
  ],
  '/mock filename with quotes ", semicolons ; and spaces.js',
];

export { postprocessArguments };
