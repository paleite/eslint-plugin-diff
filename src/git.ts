import * as child_process from "child_process";
import { resolve } from "path";
import { log } from "./logging";
import { Range } from "./Range";

const COMMAND = "git";
const OPTIONS = { maxBuffer: 1024 * 1024 * 100 };

const getDiffForFile = (
  commit: string,
  filePath: string,
  staged = false
): string => {
  log("Getting diff for file", filePath);
  const args = [
    "diff",
    "--diff-algorithm=histogram",
    "--diff-filter=ACM",
    "--find-renames=100%",
    "--no-ext-diff",
    "--relative",
    staged && "--staged",
    "--unified=0",
    commit,
    "--",
    resolve(filePath),
  ].reduce<string[]>(
    (acc, cur) => (typeof cur === "string" ? [...acc, cur] : acc),
    []
  );

  return child_process.execFileSync(COMMAND, args, OPTIONS).toString();
};

const getDiffFileList = (commit: string, staged = false): string[] => {
  log(`Getting list of files for ${staged ? "staged files" : "changed files"}`);

  const args = [
    "diff",
    "--diff-algorithm=histogram",
    "--diff-filter=ACM",
    "--find-renames=100%",
    "--name-only",
    "--no-ext-diff",
    "--relative",
    staged && "--staged",
    commit,
    "--",
  ].reduce<string[]>(
    (acc, cur) => (typeof cur === "string" ? [...acc, cur] : acc),
    []
  );

  const diffFileList = child_process
    .execFileSync(COMMAND, args, OPTIONS)
    .toString()
    .trim()
    .split("\n")
    .map((filePath) => resolve(filePath));

  log(diffFileList);

  return diffFileList;
};

const hasCleanIndex = (filePath: string): boolean => {
  log("Checking if file has clean index");

  const args = [
    "diff",
    "--no-ext-diff",
    "--quiet",
    "--relative",
    "--unified=0",
    "--",
    resolve(filePath),
  ];

  try {
    child_process.execFileSync(COMMAND, args, OPTIONS);
  } catch (err: unknown) {
    return false;
  }

  return true;
};

const fetchFromOrigin = (branch: string) => {
  log("Fetching from origin", branch);
  const args = ["fetch", "--quiet", "origin", branch];

  child_process.execFileSync(COMMAND, args, OPTIONS);
};

let untrackedFileListCache: string[] | undefined;
const getUntrackedFileList = (
  staged = false,
  shouldRefresh = false
): string[] => {
  if (staged) {
    return [];
  }

  if (untrackedFileListCache === undefined || shouldRefresh) {
    log("Getting list of files for untracked files");
    const args = ["ls-files", "--exclude-standard", "--others"];

    untrackedFileListCache = child_process
      .execFileSync(COMMAND, args, OPTIONS)
      .toString()
      .trim()
      .split("\n")
      .map((filePath) => resolve(filePath));

    log("Untracked files", untrackedFileListCache);
  }

  return untrackedFileListCache;
};

const isHunkHeader = (input: string) => {
  const hunkHeaderRE = /^@@ [^@]* @@/u;

  return hunkHeaderRE.exec(input);
};

const getRangeForChangedLines = (line: string) => {
  /**
   * Example values of the RegExp's group:
   *
   * start: '7',
   * linesCountDelimiter: ',2',
   * linesCount: '2',
   */
  const rangeRE =
    /^@@ .* \+(?<start>\d+)(?<linesCountDelimiter>,(?<linesCount>\d+))? @@/u;
  const range = rangeRE.exec(line);
  if (range === null) {
    throw Error(`Couldn't match regex with line '${line}'`);
  }

  const groups = {
    // Fallback value to ensure hasAddedLines resolves to false
    start: "0",
    linesCountDelimiter: ",0",
    linesCount: "0",
    ...range.groups,
  };

  const linesCount: number =
    groups.linesCountDelimiter && groups.linesCount
      ? parseInt(groups.linesCount)
      : 1;

  const hasAddedLines = linesCount !== 0;
  const start: number = parseInt(groups.start);
  const end = start + linesCount;

  return hasAddedLines ? new Range(start, end) : null;
};

const getRangesForDiff = (diff: string): Range[] => {
  log("Getting ranges for diff");

  return diff.split("\n").reduce<Range[]>((ranges, line) => {
    if (!isHunkHeader(line)) {
      return ranges;
    }

    const range = getRangeForChangedLines(line);
    if (range === null) {
      return ranges;
    }

    return [...ranges, range];
  }, []);
};

export {
  fetchFromOrigin,
  getDiffFileList,
  getDiffForFile,
  getRangesForDiff,
  getUntrackedFileList,
  hasCleanIndex,
};
