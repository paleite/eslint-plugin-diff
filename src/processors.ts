import type { Linter } from "eslint";
import { guessBranch } from "./ci";
import {
  fetchFromOrigin,
  getDiffFileList,
  getDiffForFile,
  getRangesForDiff,
  getUntrackedFileList,
  hasCleanIndex,
} from "./git";
import type { Range } from "./Range";

if (process.env.CI !== undefined) {
  const branch = process.env.ESLINT_PLUGIN_DIFF_COMMIT ?? guessBranch();
  if (branch !== undefined) {
    fetchFromOrigin(branch);
    process.env.ESLINT_PLUGIN_DIFF_COMMIT = branch;
  }
}

/**
 * Exclude unchanged files from being processed
 *
 * Since we're excluding unchanged files in the post-processor, we can exclude
 * them from being processed in the first place, as a performance optimization.
 * This is increasingly useful the more files there are in the repository.
 */
const getPreProcessor =
  (diffFileList: string[], staged: boolean) =>
  (text: string, filename: string) => {
    let untrackedFileList = getUntrackedFileList(staged);
    const shouldRefresh =
      !diffFileList.includes(filename) && !untrackedFileList.includes(filename);
    if (shouldRefresh) {
      untrackedFileList = getUntrackedFileList(staged, true);
    }
    const shouldBeProcessed =
      process.env.VSCODE_PID !== undefined ||
      diffFileList.includes(filename) ||
      untrackedFileList.includes(filename);

    return shouldBeProcessed ? [text] : [];
  };

const isLineWithinRange = (line: number) => (range: Range) =>
  range.isWithinRange(line);

/**
 * @internal
 */
const getUnstagedChangesError = (filename: string): [Linter.LintMessage] => {
  // When we only want to diff staged files, but the file is partially
  // staged, the ranges of the staged diff might not match the ranges of the
  // unstaged diff and could cause a conflict, so we return a fatal
  // error-message instead.

  const fatal = true;
  const message = `${filename} has unstaged changes. Please stage or remove the changes.`;
  const severity: Linter.Severity = 2;
  const fatalError: Linter.LintMessage = {
    fatal,
    message,
    severity,
    column: 0,
    line: 0,
    ruleId: null,
  };

  return [fatalError];
};

const getPostProcessor =
  (staged: boolean) =>
  (
    messages: Linter.LintMessage[][],
    filename: string
  ): Linter.LintMessage[] => {
    if (messages.length === 0) {
      // No need to filter, just return
      return [];
    }
    const untrackedFileList = getUntrackedFileList(staged);
    if (untrackedFileList.includes(filename)) {
      // We don't need to filter the messages of untracked files because they
      // would all be kept anyway, so we return them as-is.
      return messages.flat();
    }

    if (staged && !hasCleanIndex(filename)) {
      return getUnstagedChangesError(filename);
    }

    const rangesForDiff = getRangesForDiff(getDiffForFile(filename, staged));

    return messages.flatMap((message) => {
      const filteredMessage = message.filter(({ fatal, line }) => {
        if (fatal === true) {
          return true;
        }

        const isLineWithinSomeRange = rangesForDiff.some(
          isLineWithinRange(line)
        );

        return isLineWithinSomeRange;
      });

      return filteredMessage;
    });
  };

type ProcessorType = "diff" | "staged" | "ci";

const getProcessors = (
  processorType: ProcessorType
): Required<Linter.Processor> => {
  const staged = processorType === "staged";
  const diffFileList = getDiffFileList(staged);

  return {
    preprocess: getPreProcessor(diffFileList, staged),
    postprocess: getPostProcessor(staged),
    supportsAutofix: true,
  };
};

const ci = process.env.CI !== undefined ? getProcessors("ci") : {};
const diff = getProcessors("diff");
const staged = getProcessors("staged");

const diffConfig: Linter.BaseConfig = {
  plugins: ["diff"],
  overrides: [
    {
      files: ["*"],
      processor: "diff/diff",
    },
  ],
};

const ciConfig: Linter.BaseConfig =
  process.env.CI === undefined
    ? {}
    : {
        plugins: ["diff"],
        overrides: [
          {
            files: ["*"],
            processor: "diff/ci",
          },
        ],
      };

const stagedConfig: Linter.BaseConfig = {
  plugins: ["diff"],
  overrides: [
    {
      files: ["*"],
      processor: "diff/staged",
    },
  ],
};

export {
  ci,
  ciConfig,
  diff,
  diffConfig,
  staged,
  stagedConfig,
  getUnstagedChangesError,
};
