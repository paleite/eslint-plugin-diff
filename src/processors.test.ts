import * as child_process from "child_process";
import type { Linter } from "eslint";
import { mocked } from "jest-mock";
import * as git from "./git";
import { diff, diffConfig, staged, stagedConfig } from "./processors";
import {
  diff as fixtureDiff,
  staged as fixtureStaged,
} from "./__fixtures__/diff";
import { postprocessArguments } from "./__fixtures__/postprocessArguments";

jest.mock("child_process");
const mockedChildProcess = mocked(child_process, true);
const mockFilename = '/mock filename with quotes ", semicolons ; and spaces.js';
mockedChildProcess.execSync.mockReturnValue(Buffer.from(mockFilename));
const mockedGit = mocked(git, true);

// @ts-expect-error todo: fix type error here
mockedGit.getDiffFileList = jest.fn();

const [messages, filename] = postprocessArguments;

describe("processors", () => {
  const validFilename = filename;
  const sourceCode = "/** Some source code */";

  it("diff preprocess", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(fixtureDiff));

    expect(
      diff.preprocess && diff.preprocess(sourceCode, validFilename)
    ).toEqual([sourceCode]);
  });

  /**
   * Linked issue: https://github.com/paleite/eslint-plugin-diff/issues/14
   */
  it("diff preprocess should not exclude files in a Vscode plugin context", () => {
    process.env.VSCODE_CLI = "true";
    process.env.VSCODE_PID = "9293";

    mockedChildProcess.execSync.mockReturnValue(Buffer.from(""));

    expect(
      diff.preprocess && diff.preprocess(sourceCode, validFilename)
    ).toEqual([sourceCode]);

    // clean it up
    process.env.VSCODE_PID = undefined;
    process.env.VSCODE_CLI = undefined;
  });

  it("staged preprocess", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(fixtureDiff));

    expect(
      diff.preprocess && diff.preprocess(sourceCode, validFilename)
    ).toEqual([sourceCode]);
  });

  it("diff postprocess", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(fixtureDiff));

    expect(
      diff.postprocess && diff.postprocess(messages, filename)
    ).toMatchSnapshot();

    expect(mockedChildProcess.execSync).toHaveBeenCalled();
  });

  it("staged postprocess", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(fixtureStaged));

    expect(
      staged.postprocess && staged.postprocess(messages, filename)
    ).toMatchSnapshot();

    expect(mockedChildProcess.execSync).toHaveBeenCalled();
  });

  it("should report fatal errors", () => {
    mockedChildProcess.execSync.mockReturnValue(Buffer.from(fixtureDiff));

    const [[firstMessage, ...restMessage], ...restMessageArray] = messages;
    const messagesWithFatal: Linter.LintMessage[][] = [
      [{ ...firstMessage, fatal: true }, ...restMessage],
      ...restMessageArray,
    ];

    expect(
      diff.postprocess && diff.postprocess(messages, filename)
    ).toHaveLength(2);
    expect(
      diff.postprocess && diff.postprocess(messagesWithFatal, filename)
    ).toHaveLength(3);

    expect(mockedChildProcess.execSync).toHaveBeenCalled();
  });
});

describe("configs", () => {
  it("diff", () => {
    expect(diffConfig).toMatchSnapshot();
  });

  it("staged", () => {
    expect(stagedConfig).toMatchSnapshot();
  });
});
