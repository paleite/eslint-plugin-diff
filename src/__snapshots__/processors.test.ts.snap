// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configs diff 1`] = `
{
  "overrides": [
    {
      "files": [
        "*",
      ],
      "processor": "diff/diff",
    },
  ],
  "plugins": [
    "diff",
  ],
}
`;

exports[`configs staged 1`] = `
{
  "overrides": [
    {
      "files": [
        "*",
      ],
      "processor": "diff/staged",
    },
  ],
  "plugins": [
    "diff",
  ],
}
`;

exports[`processors diff postprocess 1`] = `
[
  {
    "column": 1,
    "line": 2,
    "message": "Expected { after 'if' condition.",
    "messageId": "missingCurlyAfterCondition",
    "nodeType": "IfStatement",
    "ruleId": "curly",
    "severity": 1,
  },
  {
    "column": 1,
    "fix": Object {
      "range": Array [
        0,
        3,
      ],
      "text": "",
    },
    "line": 3,
    "message": "Expected { after 'if' condition.",
    "messageId": "missingCurlyAfterCondition",
    "nodeType": "IfStatement",
    "ruleId": "curly",
    "severity": 1,
  },
]
`;

exports[`processors staged postprocess 1`] = `
[
  {
    "column": 1,
    "line": 2,
    "message": "Expected { after 'if' condition.",
    "messageId": "missingCurlyAfterCondition",
    "nodeType": "IfStatement",
    "ruleId": "curly",
    "severity": 1,
  },
  Object {
    "column": 1,
    "fix": Object {
      "range": Array [
        0,
        3,
      ],
      "text": "",
    },
    "line": 3,
    "message": "Expected { after 'if' condition.",
    "messageId": "missingCurlyAfterCondition",
    "nodeType": "IfStatement",
    "ruleId": "curly",
    "severity": 1,
  },
]
`;
